keps/sig-storage/20180515-svcacct-token-volumes.md

```diff
@@ -159,6 +162,40 @@ sources:
     audience: ca.istio.io
 ```
 
+### File Permission
+
+
+1. If there is only one UserID (non-root) shared by all the containers in a pod,
+set the owner of projected service account volume to the UserID and mode to 600.
+2. Otherwise, set the owners to be the UserID of the first container in the pod
+and the group either specified by `fsGroup` in the `PodSecurityContext` or
+unused GroupID and set the mode to 0640.
+
+UserID can be obtained by the following three ways:
+1. `runAsUser` in `SecurityContext`
+2. fetch UserID from image status
+3. 0 by default
+
+GroupID can be obtained by the following three ways:
+1. `fsGroup` in `SecurityContext`
+2. set a well-known GroupID (users can easily change it by setting `fsGroup`),
+or generate a random GroupID from set {GID_MIN ~ GID_MAX - used GroupIDs
+(`runAsGroup`, `supplementalGroups`, ..., etc.)}
+3. 0 by default
+
+we can also enforce the provisioning of UserID and GroupID in `SecurityContext`
+to minimize the changes in kubelet.
+
+#### Alternatives:
+1. same as the proposal except for creating a different volume for a different
+user ID instead of setting GroupID ownership.
+    + projected service account token is a pod level resource
+    + scalability issue. # of volumes = # of user IDs.
```

Comment:
if we reuse the same token (to avoid a renewal loop per container), is the scale we'd hit at up to one volume per container problematic?

Reply:
We have the data in cache in the kubelet so this is purely scalability of kubelet volume management. IMO the major downside of this approach is implementation complexity.

Reply:
yes, besides the implementation complexity of managing volumes for different uids, i'd also like the unix primitive idea of group.